### PR TITLE
Creating multi-nodes SFCs with xconnect and vxlans

### PR DESCRIPTION
**Pull request by AdelBouridah**
This is to make contiv/vpp working with multi-nodes SFCs over cross-connect (tap interfaces) and vxlans. vxlans refer to different VNI   to allow creating them for different SFCs. VNIs range for SFCs begin from 5000  (>=5000) to differentiate  them from other purpose vxLAN.  We note that only one SFC instance is take into account for each SFC type. multi-SFC instances will be taken into account later (load-balancing i.e. tackle the traffic streering problem).

### CODE REVIEW

File: plugins/ipam/ipam.go [hunk 1]
```diff
@@ -1061,3 +1068,98 @@ func customIfID(ifName, network string) string {
 	}
 	return ifName + "/" + network
 }
+
+////////////////////////////////////////////////////////////////////////////////
```
Comment (rastislavszabo) [inline plugins/ipam/ipam.go]: You can get rid of this comment, it does no need to be marked this way.
Feel free to modify the copyright header of the source files you touched.

File: vagrant/config/cert [hunk 2]
```diff
@@ -0,0 +1,63 @@
+[init] using Kubernetes version: v1.12.3
```
Comment (rastislavszabo) [inline vagrant/config/cert]: plase do not commit changes in the vagrant folder

File: vagrant/config/init [hunk 3]
```diff
@@ -0,0 +1,2 @@
+export KUBEADM_TOKEN=eskp57.iti9ygjwgvl3p8k2
```
Comment (rastislavszabo) [inline vagrant/config/init]: plase do not commit changes in the vagrant folder

File: plugins/ipam/ipam.go [hunk 4]
```diff
@@ -1061,3 +1068,98 @@ func customIfID(ifName, network string) string {
...
+//    Added By Adel For vni allocation of the SFC vxlan management            //
+////////////////////////////////////////////////////////////////////////////////
+
+// findFirstAvailableID returns the smallest integer that is not present in the
+// given slice of already allocated node IDs.
+func findFirstAvailableVNI(vnis []int) int {
+	res := 5000 // SFC vxlan vni begin from 5000, (low vni <5000 are for other vxlan purpose (like vni=10 for the full-mesched contiv-vpp vxlans)
```
Comment (rastislavszabo) [inline plugins/ipam/ipam.go]: please define the number 5000 as a constant

**Comment by rastislavszabo** [comment]
Please see the Travis build errors: 
https://travis-ci.org/contiv/vpp/builds/567937923?utm_source=github_status&utm_medium=notification

Please, go format the following files:
``
plugins/sfc/renderer/l2xconn/l2xconn_renderer.go.orig	2019-08-05
plugins/sfc/renderer/l2xconn/l2xconn_renderer.go	2019-08-05
``
Also please note the issues in the YAML files reported by yamllint. Are all the yaml files really useful?

File: plugins/ipam/ipam.go [hunk 5]
```diff
@@ -1061,3 +1068,98 @@ func customIfID(ifName, network string) string {
...
+
+	key := vnialloc.Key(sfcName, sfcInsNbr)
+	allocation := &vnialloc.CustomVniVxlanSFC{}
+
+	db, err := i.getDBBroker()
+	if err != nil {
+		return 0, err
+	}
+
+	// try to read existing allocation, otherwise create new
```
Comment (rastislavszabo) [inline plugins/ipam/ipam.go]: It is not a good practice to look into the database for each VNI allocation request. You should instead store the allocations in memory (e.g. in some map) and only access database in two cases:
- to persist a new allocation (see for example https://github.com/contiv/vpp/blob/dev/plugins/ipam/ipam.go#L617)
- in resync phase to populate the in-memory map with persisted data (for example see  https://github.com/contiv/vpp/blob/dev/plugins/ipam/ipam.go#L254)

File: plugins/ipam/ipam_api.go [hunk 6]
```diff
@@ -133,6 +133,9 @@ type API interface {
 	// JSON flag for every option). If contivCIDR is used it returns actual
 	// dissected subnets.
 	GetIPAMConfigForJSON() *config.IPAMConfig
+
+	// AllocateVNI tries to allocate a VNI for the SFC instance if not always allocated then persiste it
+	AllocateVNI(sfcName string, sfcInsNbr uint32) (vni uint32, err error)
```
Comment (rastislavszabo) [inline plugins/ipam/ipam_api.go]: Please, make this API more generic, not tied to SFCs. We will need to allocate VNIs not only for SFC purposes. I would like to have it like:
```go
// AllocateVxlanVNI tries to allocate a free VNI for the VXLAN with given name. 
// If the given VXLAN already has a VNI allocated, returns the existing allocation.
AllocateVxlanVNI(vxlanName string)  (vni uint32, err error)

// GetVxlanVNI returns an existing VNI allocation for the VXLAN with given name. 
// found is false if no allocation for the given VXLAN name exists.
GetVxlanVNI(vxlanName string)  (vni uint32, found bool)
```

File: plugins/ipam/vnialloc/vnialloc.proto [hunk 7]
```diff
@@ -0,0 +1,10 @@
+syntax = "proto3";
+
+package vnialloc;
+
+// CustomVniVxlanSFC represents vni allocation made for a custom vxlan made for an SFC
+message CustomVniVxlanSFC {
+    string sfc_name = 1;            // name of the service Function Chain
```
Comment (rastislavszabo) [inline plugins/ipam/vnialloc/vnialloc.proto]: As noted above, please make this more generic, not tied to SFCs, e.g.:
```proto
message VxlanVniAllocation {
    string vxlan_name= 1;   
    uint32 vni = 2;  
}
```

File: plugins/sfc/renderer/l2xconn/l2xconn_renderer.go [hunk 8]
```diff
@@ -69,9 +70,9 @@ func (rndr *Renderer) AfterInit() error {
 
 // AddChain is called for a newly added service function chain.
 func (rndr *Renderer) AddChain(sfc *renderer.ContivSFC) error {
-	rndr.Log.Infof("Add SFC: %v", sfc)
+	rndr.Log.Infof("By Adel 1 - Add SFC: %v AfterSFC", sfc)
```
Comment (rastislavszabo) [inline plugins/sfc/renderer/l2xconn/l2xconn_renderer.go]: Please get rid of the "By Adel" part in the log

File: plugins/sfc/renderer/l2xconn/l2xconn_renderer.go [hunk 9]
```diff
@@ -69,9 +70,9 @@ func (rndr *Renderer) AfterInit() error {
...
 
-	txn := rndr.UpdateTxnFactory(fmt.Sprintf("add SFC '%s'", sfc.Name))
+	txn := rndr.UpdateTxnFactory(fmt.Sprintf("By Adel 2 - add SFC '%s'", sfc.Name))
```
Comment (rastislavszabo) [inline plugins/sfc/renderer/l2xconn/l2xconn_renderer.go]: Please get rid of the "By Adel" part in the log

File: plugins/sfc/renderer/l2xconn/l2xconn_renderer.go [hunk 10]
```diff
@@ -187,3 +217,126 @@ func (rndr *Renderer) getSFInterface(sf *renderer.ServiceFunction, input bool) s
...
+		return 0
+	}
+	return 0
+
+}
+
+// Check weather the given sf is in the local node or not
+func (rndr *Renderer) checkSFNodeLocal(sf *renderer.ServiceFunction) bool {
+	if sf.Type != renderer.Pod {
+		return false // TODO: to be checked if is it possible:  no render for the SFC
```
Comment (rastislavszabo) [inline plugins/sfc/renderer/l2xconn/l2xconn_renderer.go]: It is possible - Interface SF has the `local` flag:

https://github.com/contiv/vpp/blob/dev/plugins/sfc/renderer/api.go#L165-L166

File: plugins/sfc/renderer/l2xconn/l2xconn_renderer.go [hunk 11]
```diff
@@ -129,35 +130,64 @@ func (rndr *Renderer) Close() error {
...
+						rndr.Log.Infof("Unable to allocate SFC VNI: %v", err)
+						break
+					}
+					if rndr.checkSFNodeLocal(sf) { // sf is local
+						// create vxlan and connect sf to vxlan in both directions
+						// direction 1
+						vxlanName, vxlanConfig := rndr.vxlanSFC(sf, prevSF, vni)
+						rndr.mergeConfiguration(config, vxlanConfig)
+
+						xconnect := rndr.connectIfaces(iface, vxlanName)
```
Comment (rastislavszabo) [inline plugins/sfc/renderer/l2xconn/l2xconn_renderer.go]: why are you not using `crossConnectIfaces` here?

File: plugins/sfc/renderer/l2xconn/l2xconn_renderer.go [hunk 12]
```diff
@@ -129,35 +130,64 @@ func (rndr *Renderer) Close() error {
...
+						// direction 2
+						xconnect = rndr.connectIfaces(vxlanName, iface)
+						rndr.mergeConfiguration(config, xconnect)
+					} else { // prevSF is local
+						// create vxlan and connect prevSF to vxlan in both directions
+						vxlanName, vxlanConfig := rndr.vxlanSFC(prevSF, sf, vni)
+						rndr.mergeConfiguration(config, vxlanConfig)
+
+						//direction 1
+						xconnect := rndr.connectIfaces(vxlanName, prevIface)
```
Comment (rastislavszabo) [inline plugins/sfc/renderer/l2xconn/l2xconn_renderer.go]: why are you not using `crossConnectIfaces` here?

File: plugins/ipam/ipam.go (hunk 1 above)
Comment (AdelBouridah) [reply plugins/ipam/ipam.go]: I have modified the copyright header. Feel free to review if it is correct?

File: vagrant/config/cert (hunk 2 above)
Comment (AdelBouridah) [reply vagrant/config/cert]: it is Ok I have added the vagrant folder to the .gitignore

File: vagrant/config/init (hunk 3 above)
Comment (AdelBouridah) [reply vagrant/config/init]: it is Ok I have added the vagrant folder to the .gitignore

File: plugins/ipam/ipam.go (hunk 5 above)
Comment (AdelBouridah) [reply plugins/ipam/ipam.go]: It is ok. I have created a map (assignedVxlanVNIs map[string]vxlanVniAllocation) that will contain the VNI allocations in memory and resync it from persisted data.

File: plugins/ipam/ipam_api.go (hunk 6 above)
Comment (AdelBouridah) [reply plugins/ipam/ipam_api.go]: done as indicated. Check please if it correspond to what demanded.

File: plugins/ipam/vnialloc/vnialloc.proto (hunk 7 above)
Comment (AdelBouridah) [reply plugins/ipam/vnialloc/vnialloc.proto]: done.

File: plugins/sfc/renderer/l2xconn/l2xconn_renderer.go (hunk 8 above)
Comment (AdelBouridah) [reply plugins/sfc/renderer/l2xconn/l2xconn_renderer.go]: I have forget these local work comments on the committed version. I have deleted all personnel dev logs.

File: plugins/sfc/renderer/l2xconn/l2xconn_renderer.go (hunk 9 above)
Comment (AdelBouridah) [reply plugins/sfc/renderer/l2xconn/l2xconn_renderer.go]: I have deleted all personnel dev logs

File: plugins/sfc/renderer/l2xconn/l2xconn_renderer.go (hunk 10 above)
Comment (AdelBouridah) [reply plugins/sfc/renderer/l2xconn/l2xconn_renderer.go]: The SF could be  a pod or an ExternalInterfaces. For Pod we have the local flag but for ExternalInterfaces we have to check if there is a local ExternalInterface or not --> I have changed the code of this Function in the new PR.

File: plugins/sfc/renderer/l2xconn/l2xconn_renderer.go (hunk 11 above)
Comment (AdelBouridah) [reply plugins/sfc/renderer/l2xconn/l2xconn_renderer.go]: Ops yeah it is possible to use the crossConnectIfaces. I have mismatched the code at the beginning (by doing a lot of change here).  do you want that I redo it with the crossConnectIfaces in a new PR?

File: plugins/sfc/renderer/l2xconn/l2xconn_renderer.go (hunk 12 above)
Comment (AdelBouridah) [reply plugins/sfc/renderer/l2xconn/l2xconn_renderer.go]: Same response for the above one

**Pull request merged**